clients/application-kafka/src/main/java/org/eclipse/hono/application/client/kafka/impl/DownstreamMessageConsumer.java

```diff
@@ -0,0 +1,119 @@
+import org.eclipse.hono.application.client.kafka.KafkaMessageContext;
+import org.eclipse.hono.application.client.kafka.KafkaMessageProperties;
+import org.eclipse.hono.kafka.client.HonoTopic;
+import org.eclipse.hono.kafka.client.KafkaMessageHelper;
+import org.eclipse.hono.kafka.client.consumer.AbstractAtLeastOnceKafkaConsumer;
+import org.eclipse.hono.util.MessageHelper;
+import org.eclipse.hono.util.QoS;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.kafka.client.consumer.KafkaConsumer;
+import io.vertx.kafka.client.consumer.KafkaConsumerRecord;
+import io.vertx.kafka.client.producer.KafkaHeader;
+
+/**
+ * A consumer that consumes downstream messages from Kafka.
+ * <p>
+ * This consumer needs to be closed by invoking {@link #close()} ()} when it is no longer needed.
+ */
+public abstract class DownstreamMessageConsumer extends AbstractAtLeastOnceKafkaConsumer<DownstreamMessage>
+        implements MessageConsumer<DownstreamMessage> {
+
+    /**
+     * A logger to be shared with subclasses.
+     */
+    protected final Logger log = LoggerFactory.getLogger(getClass());
+
+    /**
+     * Creates a new downstream consumer.
+     *
+     * @param kafkaConsumer The Kafka consumer to be exclusively used by this instance to consume records.
+     * @param topic The Kafka topic to consume records from.
+     * @param messageHandler The function to be invoked for each message created from a record.
+     * @param closeHandler The handler to be invoked when the Kafka consumer has been closed due to an error.
+     * @param pollTimeout The maximal number of milliseconds to wait for messages during a poll operation.
+     * @throws NullPointerException if any of the parameters is {@code null}.
+     */
+    protected DownstreamMessageConsumer(final KafkaConsumer<String, Buffer> kafkaConsumer, final String topic,
+            final Consumer<DownstreamMessage> messageHandler, final Handler<Throwable> closeHandler,
+            final long pollTimeout) {
+        super(kafkaConsumer, topic, messageHandler, closeHandler, pollTimeout);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected DownstreamMessage createMessage(final KafkaConsumerRecord<String, Buffer> record) {
+        Objects.requireNonNull(record);
+
+        final String tenantId = getTenantId(record);
+        final String deviceId = record.key();
+        final MessageProperties properties = new KafkaMessageProperties(record);
+        final String contentType = getContentType(record.headers());
+        final MessageContext messageContext = new KafkaMessageContext(record);
+        final QoS qos = getQoS(record.headers());
+        final Buffer payload = record.value();
+
+        return new DownstreamMessageImpl(tenantId, deviceId, properties, contentType, messageContext, qos, payload);
```

Comment:
IMHO it would improve encapsulation if you created a kafka specific DownstreamMessage which simply wrapped a KafkaConsumerRecord, thus hiding all the code for retrieving the property values within it ...
The DownstreamMessageImpl class can then be removed FMPOV.

Reply:
Yes, makes sense.